Toy2/src/Calculator.java

```diff
@@ -0,0 +1,209 @@
+
+    public Calculator(String input) {
+        InputParser parser = new InputParser(input);
+        this.stringInput = parser.getParsedResult();
+        validateState();
+        initializeOperations();
+    }
+    // 연산자와 그에 대응하는 연산 클래스를 매핑
+    private void initializeOperations() {
+
+
+        //ENUM을 사용하면 궁극적인 형태는??이렇게?? -> operations.put(MathSign.PLUS.getValue(), MathSign.PLUS);
+        operations.put(MathSign.PLUS.getValue(), new Plus());
+        operations.put(MathSign.MINUS.getValue(), new Minus());
+        operations.put(MathSign.MULTIPLY.getValue(), new Multiple());
+        operations.put(MathSign.DIVIDE.getValue(), new Divide());
+        operations.put(MathSign.REMAINDER.getValue(), new Remainder()); // 나머지 연산 추가
+        // 필요한 경우 다른 연산자들도 이곳에 추가 가능
+    }
+
+    //연산 가능여부의 검증
+    private void validateState() {
+        long parenthesisBalance = 0; // 괄호의 균형을 검증하기 위한 변수
+        boolean lastWasOperator = true; // 마지막 토큰이 연산자인지 검증하기 위한 변수
+
+        //for문을 이용한부분을 메소드로 뺴고싶은데 지역변수가 존재하여 여기가 최선일것같다 다른방법이있을까?
+        for (int i = 0; i < stringInput.size(); i++) {
+
+            String token = stringInput.get(i);
+
+            if (token.equals("(")) {
+                parenthesisBalance++;
+                lastWasOperator = true;
+            }
+            if (token.equals(")")) {
+                if (parenthesisBalance == 0) {
+                    throw new IllegalArgumentException("괄호 불일치 오류: 위치 " + i);
+                }
+                parenthesisBalance--;
+                lastWasOperator = false;
+            }
+            if (isNumber(token)) {
+                if (!lastWasOperator && (i > 0 && !stringInput.get(i - 1).equals("("))) {
+                    throw new IllegalArgumentException("위치 " + i + "에서 숫자가 연속됩니다: " + token);
+                }
+                lastWasOperator = false;
+            }
+            if (isOperator(token)) {
+                if (lastWasOperator && !token.equals("(")) {
+                    throw new IllegalArgumentException("위치 " + i + "에서 연산자가 연속됩니다: " + token);
+                }
+                lastWasOperator = true;
+            }
+            throw new IllegalArgumentException("위치 " + i + "에서 유효하지 않은 토큰: " + token);
+
+        }
+
+        if (parenthesisBalance != 0) {
+            throw new IllegalArgumentException("괄호 불일치 오류: 전체 수식에서 괄호가 일치하지 않습니다.");
+        }
+
+        if (lastWasOperator) {
+            throw new IllegalArgumentException("수식이 잘못되었습니다. 마지막 항목이 연산자입니다.");
+        }
+    }
+
+
+
+    //숫자 검증을 위한 메소드
+    private boolean isNumber(String str) {
+        try {
+            new BigDecimal(str);
+            return true;
+        } catch (NumberFormatException e) {
+            return false;
+        }
+    }
+
+    //연산자 검증을 위한 메소드
+    private boolean isOperator(String str) {
+        return str.equals("+") || str.equals("-") || str.equals("*") || str.equals("/") || str.equals("%");
+    }
```

Comment:
Enum을 쓰셨다면, 여기서도 사용할 수 있습니다.
```suggestion
    //연산자 검증을 위한 메소드
    private boolean isOperator(String str) {
        return operations.containsKey(str);
    }
```